firedrake/preconditioners/pmg.py

```diff
@@ -213,78 +171,60 @@ def coarsen(self, fdm, comm):
         fu = fproblem.u
         cu = firedrake.Function(cV)
 
-        def coarsen_quadrature(df, Nf, Nc):
-            # Coarsen the quadrature degree in a dictionary
-            # such that the ratio of quadrature nodes to interpolation nodes (Nq+1)/(Nf+1) is preserved
-            if isinstance(df, dict):
-                Nq = df.get("quadrature_degree", None)
-                if Nq is not None:
-                    dc = dict(df)
-                    dc["quadrature_degree"] = max(2*Nc+1, ((Nq+1) * (Nc+1) + Nf) // (Nf+1) - 1)
-                    return dc
-            return df
-
-        def coarsen_form(form, Nf, Nc, replace_d):
-            # Coarsen a form, by replacing the solution, test and trial functions, and
-            # reconstructing each integral with a coarsened quadrature degree.
-            # If form is not a Form, then return form.
-            return Form([f.reconstruct(metadata=coarsen_quadrature(f.metadata(), Nf, Nc))
-                         for f in replace(form, replace_d).integrals()]) if isinstance(form, Form) else form
-
-        def coarsen_bcs(fbcs):
-            cbcs = []
-            for bc in fbcs:
-                cV_ = cV
-                for index in bc._indices:
-                    cV_ = cV_.sub(index)
-                cbc_value = self.coarsen_bc_value(bc, cV_)
-                if type(bc) == firedrake.DirichletBC:
-                    cbcs.append(firedrake.DirichletBC(cV_, cbc_value,
-                                                      bc.sub_domain))
-                else:
-                    raise NotImplementedError("Unsupported BC type, please get in touch if you need this")
-            return cbcs
-
-        Nf = PMGBase.max_degree(fV.ufl_element())
-        Nc = PMGBase.max_degree(cV.ufl_element())
-
-        # Replace dictionary with coarse state, test and trial functions
-        replace_d = {fu: cu,
-                     test: firedrake.TestFunction(cV),
-                     trial: firedrake.TrialFunction(cV)}
-
-        cF = coarsen_form(fctx.F, Nf, Nc, replace_d)
-        cJ = coarsen_form(fctx.J, Nf, Nc, replace_d)
-        cJp = coarsen_form(fctx.Jp, Nf, Nc, replace_d)
-        fcp = coarsen_quadrature(fproblem.form_compiler_parameters, Nf, Nc)
-        cbcs = coarsen_bcs(fproblem.bcs)
+        fdeg = PMGBase.max_degree(fV.ufl_element())
+        cdeg = PMGBase.max_degree(cV.ufl_element())
+
+        fine_to_coarse_map = {fu: cu,
+                              test: test.reconstruct(function_space=cV),
+                              trial: trial.reconstruct(function_space=cV)}
+
+        def _coarsen_form(a):
+            if isinstance(a, ufl.Form):
+                a = self.coarsen_form(a, fine_to_coarse_map)
+                a = type(a)([f.reconstruct(metadata=self.coarsen_quadrature(f.metadata(), fdeg, cdeg))
+                             for f in a.integrals()])
+            return a
```

Comment:
I guess this is handling the case where one of `F`, `J` and `Jp` are `None`?

Why does `coarsen_form` not handle the quadrature coarsening?